OpenSim/Actuators/Thelen2003Muscle.cpp

```diff
@@ -762,110 +764,163 @@ Thelen2003Muscle::initMuscleState(const SimTK::State& s,
+                // We've stagnated or hit a limit; assume we are hitting local
+                // minimum, attempt to jump to the other side of the force-length
+                // curve over the optimal fiber-length assuming we're stuck on one
+                // side of the curve.
+                lce = lcePrev - sign(delta_lce)*SimTK::SqrtEps;
+                h = 0;
+            }
+
+            if (lce < getMinimumFiberLength()) {
+                lce = 0.1*getOptimalFiberLength();
+                h = 0; // force a break
+            }
+
+            //Update position level quantities, only if they won't go 
+            //singular
+            phi = getPennationModel().calcPennationAngle(lce);
+            cosphi = cos(phi);
+            tl = ml - lce*cosphi;
+            lceN = lce / ofl;
+            tlN = tl / tsl;
+
+            //Update the multipliers and their partial derivatives
+            fse = calcfse(tlN);
+            fal = calcfal(lceN);
+            fpe = calcfpe(lceN);
+
+            // Compute the force error assuming fiber-velocity estimated is
+            // unchanged
+            Fm = (ma*fal*fv + fpe)*fiso;
             FmAT = Fm*cosphi;
             Ft = fse*fiso;
-            ferr  = FmAT-Ft; 
-        
-            //Compute the partial derivative of the force error w.r.t. lce
-            dFm_dlce   = calcDFmDlce(lce, ma, fv, fiso, ofl);
-            dFmAT_dlce = calcDFmATDlce(lce,phi,cosphi,Fm, dFm_dlce, vol);
-            dFmAT_dlceAT= dFmAT_dlce*cosphi;
+            ferr = FmAT - Ft;
 
-            dFt_d_tl    = calcDFseDtl(tl, fiso, tsl);
-            dFt_d_lce   = calcDFseDlce(tl, lce, phi, cosphi,fiso, tsl, vol);
-            
+            if (h <= SimTK::SqrtEps ) {
+                break;
+            }
+            else
+                h = 0.5*h;;
+        }
 
-            dferr_d_lce = dFmAT_dlce - dFt_d_lce;
-
-            if(abs(ferr) > aSolTolerance){
-                //Take a full Newton Step                            
-                if(abs(dferr_d_lce) > SimTK::SignificantReal){
-                    delta_lce   = - ferr/dferr_d_lce;
-                    lce         = lce + delta_lce;                
-                }else{ //If we've stagnated, perturb the current solution            
-                    double perturbation = 
-                    2.0*((double)rand())/((double)RAND_MAX)-1.0;
-                    double lengthPerturbation = 
-                        0.5*perturbation*getOptimalFiberLength();
-                    lce = lce + lengthPerturbation;
-                }
+        ferrPrev = ferr;
+        lcePrev = lce;
+
+        //Update the partial derivative of the force error w.r.t. lce
+        dFm_dlce = calcDFmDlce(lce, ma, fv, fiso, ofl);
+        dFmAT_dlce = calcDFmATDlce(lce, phi, cosphi, Fm, dFm_dlce, vol);
+        dFmAT_dlceAT = dFmAT_dlce*cosphi;
+
+        dFt_d_tl = calcDFseDtl(tl, fiso, tsl);
+        dFt_d_lce = calcDFseDlce(tl, lce, phi, cosphi, fiso, tsl, vol);
```

Comment:
lambda functions are now employed and the algorithms between the `Thelen2003Muscle` and `Millard2012EquilibriumMuscle` are now identical and much easier to read (with slightly different local lambdas). Thanks for recommending the change.